Libraries/LibWeb/HTML/HTMLElement.cpp

```diff
@@ -1809,6 +1838,107 @@ void HTMLElement::queue_a_popover_toggle_event_task(String old_state, String new
+    auto topmost_popover = document.topmost_auto_or_hint_popover();
+
+    // 5. If topmostPopover is null, then return.
+    if (!topmost_popover)
+        return;
+
+    // 6. If event's type is "pointerdown", then: set document's popover pointerdown target to the result of running topmost clicked popover given target.
+    if (event.type() == UIEvents::EventNames::pointerdown)
+        document.set_popover_pointerdown_target(topmost_clicked_popover(target));
+
+    // 7. If event's type is "pointerup", then:
+    if (event.type() == UIEvents::EventNames::pointerup) {
+        // 1. Let ancestor be the result of running topmost clicked popover given target.
+        auto const ancestor = topmost_clicked_popover(target);
+
+        // 2. Let sameTarget be true if ancestor is document's popover pointerdown target.
+        bool const same_target = ancestor == document.popover_pointerdown_target();
+
+        // 3. Set document's popover pointerdown target to null.
+        document.set_popover_pointerdown_target({});
+
+        // 4. If ancestor is null, then set ancestor to document.
+        Variant<GC::Ptr<HTMLElement>, GC::Ptr<DOM::Document>> ancestor_or_document = ancestor;
+        if (!ancestor)
+            ancestor_or_document = GC::Ptr(document);
+
+        // 5. If sameTarget is true, then run hide all popovers until given ancestor, false, and true.
+        if (same_target)
+            hide_all_popovers_until(ancestor_or_document, FocusPreviousElement::No, FireEvents::Yes);
+    }
+}
+
+// https://html.spec.whatwg.org/multipage/popover.html#get-the-popover-stack-position
+size_t HTMLElement::popover_stack_position()
+{
+    // To get the popover stack position, given an HTML element popover:
+
+    // 1. Let hintList be popover's node document's showing hint popover list.
+    auto hint_list = document().showing_hint_popover_list();
+
+    // 2. Let autoList be popover's node document's showing auto popover list.
+    auto auto_list = document().showing_auto_popover_list();
+
+    // 3. If popover is in hintList, then return the index of popover in hintList + the size of autoList + 1.
+    if (hint_list.contains_slow(GC::Ref(*this)))
+        return hint_list.find_first_index(GC::Ref(*this)).value() + auto_list.size() + 1;
+
+    // 4. If popover is in autoList, then return the index of popover in autoList + 1.
+    if (auto_list.contains_slow(GC::Ref(*this)))
+        return auto_list.find_first_index(GC::Ref(*this)).value() + 1;
+
+    // 5. Return 0.
+    return 0;
+}
+
+// https://html.spec.whatwg.org/multipage/popover.html#topmost-clicked-popover
+GC::Ptr<HTMLElement> HTMLElement::topmost_clicked_popover(GC::Ptr<DOM::Node> node)
+{
+    // To find the topmost clicked popover, given a Node node:
+
+    GC::Ptr<HTMLElement> nearest_element = as_if<HTMLElement>(*node);
+    if (!nearest_element)
+        nearest_element = node->shadow_including_first_ancestor_of_type<HTMLElement>();
+
+    if (!nearest_element)
+        return {};
```

Comment:
It feels a little sketchy to have to do gymnastics like this to call a method that the spec says applies to a `Node` anyway. Should `nearest_inclusive_open_popover()` be a method of `DOM::Node` instead of `HTMLElement`? A few other methods there also refer to being given a `Node`.

Reply:
Yeah, the main reason is that the popover attribute and visibility state are only defined for html elements, so those gymnastics would always need to be done _somewhere_, and I thought it would be better to keep the popover implementation as localized as possible.

Reply:
Flicking through the spec, it seems to only ever ask an HTMLElement for its visibility state, so that should be alright... I do see it asking for the `popovertargetaction` attribute on a Node though, yeah. Not sure what the correct way is of handling asking for an attribute's state where the attribute isn't legal.

It's probably fine for now.

Reply:
Oh the `popover` attribute gets used a lot, I see. Same issue though. :thonk:

Reply:
yeah, not sure if it counts as a spec bug or if we're supposed to assume that "not defined" -> "not present" -> "!= auto or hint"

Reply:
I've asked for clarification in the whatwg matrix, but for now do you want me to make any changes?